core/mount/mount_linux_test.go

```diff
@@ -198,6 +206,120 @@ func TestUnmountRecursive(t *testing.T) {
 	}
 }
 
+func TestDoPrepareIDMappedOverlay(t *testing.T) {
+	testutil.RequiresRoot(t)
+
+	k512 := kernel.KernelVersion{Kernel: 5, Major: 12}
+	ok, err := kernel.GreaterEqualThan(k512)
+	require.NoError(t, err)
+	if !ok {
+		t.Skip("GetUsernsFD requires kernel >= 5.12")
+	}
+
+	usernsFD, err := getUsernsFD(testUIDMaps, testGIDMaps)
+	require.NoError(t, err)
+	defer usernsFD.Close()
+
+	type testCase struct {
+		name              string
+		injectUmountFault bool
+	}
+
+	tcases := []testCase{
+		{
+			name:              "normal",
+			injectUmountFault: false,
+		},
+		{
+			name:              "umount-fault",
+			injectUmountFault: true,
+		},
+	}
+
+	for _, tc := range tcases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Reserve an OS thread with full capabilities for cleanup.
+			testCompleteChan := make(chan struct{})
+			cleanup := func() {}
+			runtime.LockOSThread()
+			defer runtime.UnlockOSThread()
+
+			// Force the go runtime to run the actual test on a separate thread, where we may drop caps.
+			go func() {
+				defer close(testCompleteChan)
+				fakeLowerDirsDir := t.TempDir()
+				lowerDirs := []string{filepath.Join(fakeLowerDirsDir, "lower1"), filepath.Join(fakeLowerDirsDir, "lower2")}
+				for _, dir := range lowerDirs {
+					require.NoError(t, os.Mkdir(dir, 0755))
+					require.NoError(t, os.WriteFile(filepath.Join(dir, filepath.Base(dir)), []byte("foo"), 0644))
+				}
+
+				remountsLocation := t.TempDir()
+
+				tmpLowerDirs, c, err := doPrepareIDMappedOverlay(remountsLocation, lowerDirs, int(usernsFD.Fd()))
+				cleanup = c
+				require.NoError(t, err)
+				require.Len(t, tmpLowerDirs, len(lowerDirs))
+
+				lowerContents := make([][]byte, len(lowerDirs))
+
+				for i, dir := range lowerDirs {
+					correspondingRemount := tmpLowerDirs[i]
+					filename := filepath.Base(dir)
+
+					expectedFile, err := os.ReadFile(filepath.Join(dir, filename))
+					require.NoError(t, err, "reading comparison test fixture file")
+					lowerContents[i] = expectedFile
+
+					actualFile, err := os.ReadFile(filepath.Join(correspondingRemount, filename))
+					require.NoError(t, err, "reading file in temporary remount")
+
+					assert.Equal(t, expectedFile, actualFile, "file content in temporary remount")
+				}
+
+				if tc.injectUmountFault {
+					runtime.LockOSThread()
+					// Never unlock in this goroutine, ensures the thread isn't reused.
+					// Thus we don't have to worry about restoring the caps.
+					ourCaps, err := capability.NewPid2(0)
```

Comment:
is it thread-leader level or sub thread level?

Reply:
In practice it's coming up as not the group leader but the reason why is perhaps fragile:
* At the test function level eg `TestdoPrepareIdMappedOverlay`, `go test` runs this function on the leader thread.
* Each `t.Run()` is in a new goroutine but because we have not `t.Parallel()`ized there is no need to spawn new threads and the go runtime is mapping the new goroutines to the leader thread. Here we call `runtime.LockOSThread()` the first time.
* Then we spawn a new goroutine in which we conduct our real business. At this point the runtime is required to create a new thread since it previously only had the leader which is now locked. This new thread must not be the leader.

We'll see if we keep this, as @rata also points out it _is_ pretty complicated.

Reply:
Good to know this. Since you prepare tempdir for temporary mountpoint, we can try to open file in that mountpoint, which can cause EBUSY during umount. Maybe it's easier to reproduce it this issue instead of using `runtime.LockOSThread()`. What do you think? :)

Reply:
If not using MNT_DETACH yes that works and is easier.